File: chain/golang/boiling_chef.go
package main

import "fmt"

type BoilingChef struct {
	next Chef
}

func (b *BoilingChef) Execute(d *Dish) {
	fmt.Printf("Boiling %s...\n", d.Name)
	d.IsBoiled = true
	if b.next != nil {
		b.next.Execute(d)
	}
}

func (b *BoilingChef) SetNext(next Chef) {
	b.next = next
}

File: chain/rust/chain/src/boiling_chef.rs
use crate::chef::Chef;
use crate::basic_chef::BasicChef;

pub struct BoilingChef {
    pub base: BasicChef,
}

impl Chef for BoilingChef {
    fn set_next(&mut self, next: Box<dyn Chef>) { self.base.next = Some(next); }
    fn cook(&self, dish: &str) {
        if dish.contains("Soup") {
            println!("Boiling the soup...");
        }
        if let Some(ref next) = self.base.next { next.cook(dish); }
    }
}
